### PR TITLE
Added additional scanner cancel points ..

### PR DESCRIPTION
**Pull request by daschuer**
... during the library clean up time. 

This is required, since the clean up time task may take significant > 20 s time, which may delay the Mixxx shutdown. 

Breaking the clean up should be save, since we use a transaction. 
A "cancel" during the clean up period will have the same effect as a "cancel" before this region.

In the original code the clean up tasks where skipped by a "cancel", but the found tracks where stored. 
This means with and without the patch we are facing situations where the library clean up is not run.
I think this is OK. but if not, we have to fix the upstream as well

### CODE REVIEW

**Comment by daschuer** [comment]
@kain88-de: It would be nice if you can verify it. Thank you.

File: src/library/scanner/libraryscanner.cpp [hunk 1]
```diff
@@ -251,6 +251,81 @@ void LibraryScanner::slotStartScan() {
...
+    }
+
+    qDebug() << "Marking unverified tracks as deleted.";
+    m_trackDao.markUnverifiedTracksAsDeleted();
+
+    qDebug() << "Marking unverified directories as deleted.";
+    m_libraryHashDao.markUnverifiedDirectoriesAsDeleted();
+
+    // For making the scanner slow
+   qDebug() << "Burn CPU";
```
Comment (kain88-de) [inline src/library/scanner/libraryscanner.cpp]: why is this needed?

File: src/library/scanner/libraryscanner.cpp [hunk 2]
```diff
@@ -251,6 +251,81 @@ void LibraryScanner::slotStartScan() {
...
+   for (int i = 0;i < 1000000000; i++) asm("nop");
+
+
+
+    // Check to see if the "deleted" tracks showed up in another location,
+    // and if so, do some magic to update all our tables.
+    qDebug() << "Detecting moved files.";
```
Comment (kain88-de) [inline src/library/scanner/libraryscanner.cpp]: Why would you do that after marking things as deleted? Logically it makes for sense for me to first look for moved files and do our things. Only after all of that is done check if tracks should be marked as deleted

File: src/library/dao/trackdao.cpp [hunk 3]
```diff
@@ -1622,14 +1622,16 @@ void TrackDAO::markTrackLocationsAsDeleted(const QString& directory) {
...
-    //This function should not start a transaction on it's own!
-    //When it's called from libraryscanner.cpp, there already is a transaction
-    //started!
+// Look for moved files. Look for files that have been marked as
+// "deleted on disk" and see if another "file" with the same name and
+// files size exists in the track_locations table. That means the file has
+// moved instead of being deleted outright, and so we can salvage your
+// existing metadata that you have in your DB (like cue points, etc.).
+bool TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld,
+        QSet<int>* pTracksMovedSetNew, volatile const bool* pCancel) {
```
Comment (kain88-de) [inline src/library/dao/trackdao.cpp]: why would we need to cancel this one. As I remember this operates only on the database, which should be a matter of seconds tops even on old PC's

**Comment by kain88-de** [comment]
@daschuer 

When you make claims about speed if would be nice to see some benchmarks. This is hard to time here admittedly but it should still be possible to run it in valgrind. (which might actually help since it slows things down). 

I personally haven't encountered a clean-up phase of several seconds. Do you know where most of that time is spend and under what conditions it is triggered?

File: src/library/scanner/libraryscanner.cpp (hunk 1 above)
Comment (daschuer) [inline src/library/scanner/libraryscanner.cpp]: This was only a test, I have missed to disable.

File: src/library/scanner/libraryscanner.cpp (hunk 2 above)
Comment (daschuer) [inline src/library/scanner/libraryscanner.cpp]: the detect moved function acts on the files that are marked as deleted. 
So it is required to do the marking first. 

This PR moves the cleanUpScan part to a new function to be able to cancel it without touching the way how. 

There is definitely an issue how it works, because we have a user complaining about undetected doublets. 
https://bugs.launchpad.net/mixxx/+bug/1480791
This should be issued in a separate PR, to keep this less complex.

**Comment by daschuer** [comment]
Thank you for looking into that.

> When you make claims about speed if would be nice to see some benchmarks.

Unfortunately I have currently no access to my low end device to generate test data and verify the 20 s. I will do that as soon as I can.  

But IMHO this issue is not that relevant to absolute timing values, since It fixes just the issue that the last part of the scanner is not cancel-able. We do sql queries in loops where we do not know the absolute count. 
The actual count can be very high, if a for example scan his library without an external track source mounted.  This PR adds cancel conditions to those loops.  

For me a remaining question is, if it will mess up the library some how. IMHO it does not (not worse then 1.12 in respect to https://bugs.launchpad.net/mixxx/+bug/1480791)

File: src/library/dao/trackdao.cpp [hunk 4]
```diff
@@ -1819,7 +1825,7 @@ bool TrackDAO::isTrackFormatSupported(TrackInfoObject* pTrack) const {
     return false;
 }
 
-void TrackDAO::verifyRemainingTracks() {
+bool TrackDAO::verifyRemainingTracks(volatile const bool* pCancel) {
```
Comment (kain88-de) [inline src/library/dao/trackdao.cpp]: This is the only function I see that might cause a slow down because we are doing actual IO. Here I'm ok issuing a cancel. With the others I don't see the need

**Comment by kain88-de** [comment]
Can't we reduce the number of loop by using SQLite queries? Looks like groupby could help us getting all duplicate entries at once. 

http://www.tech-recipes.com/rx/49959/finding-duplicate-records-using-group-by-in-sql-server/

In my experience SQLite queries are a matter of milliseconds.

File: src/library/dao/trackdao.cpp (hunk 3 above)
Comment (daschuer) [inline src/library/dao/trackdao.cpp]: Unfortunately we have here two encapsulated while loops using query and query2 which makes even a very short time for a single query worse.
The outer while loop takes about 0.2 ms on my recent device with only 15 tracks in library. The inner while loop takes 0.0008 ms for each track in library.  
If we assume to move a the root folder of a not uncommon 10,000 tracks query and query2 will return 10,000 entries. 
Unfortunately we compare each by each and have to do 10,000 x 10,000 compares. 
The resulting rough timing calculations looks like that:  
(0.0008 ms * 10,000 + 0.2 ms) * 10,000 = 82,000 ms 

Unfortunately I cannot backup this calculation by real data by now, but even if I have an error of x 10 it is IMHO worth to keep the cancel since it is measured on a recent device. 

Of cause we should improve this duplicate detection algorithm, but that is IMHO not a 1.12 task since it involves some regression risk. The check for cancel takes 0,0015 ms in the outer loop during my test which adds a 
penalty of 15 ms to a 10,000 track scan -> + 0.02 %

**Comment by daschuer** [comment]
Yes, there seems to be a lot room for improvements. But IMHO that is not a 1.12 task.
(except if we find the cause for https://bugs.launchpad.net/mixxx/+bug/1480791)

The real 1.12 issue I target here is that shutting down Mixxx with a running scan may delay the final shot down for a long time. This time the audiance might have no music since the resources are still blocked which prevents the start of a new Mixxx instance.

Does this suit for you?

File: src/library/dao/trackdao.cpp (hunk 3 above)
Comment (kain88-de) [inline src/library/dao/trackdao.cpp]: But then we should rather improve this function. Currently canceling means that only minimal progress is lost.
If this is canceled after 50 seconds run time in your scenario all of the progress we made so far will be lost. 
Coming to think of it. I also would expect that a canceled scan does a proper clean up. 

Your solution might fix the long shutdown time of mixxx in canceled scans but I don't think some one would rescan his entire library during a performance. The better solution in the long run is to improve the performance of this.

It should also be possible to write a unit-test for the clean-up functions. They only depend on internal library state.

Comment (daschuer) [inline src/library/dao/trackdao.cpp]: > But then we should rather improve this function.

Yes of cause. Since I have not a good and easy receipted yet we should postpone it after 1.12. 

During this PR I have tried careful not to change what happens by a cancel. It should be exactly the same as in current 1.12, except that cancelling during the clean up phase it excepted an the result should be the same as if you have cancelled the scan just before. 

> Currently canceling means that only minimal progress is lost.

These are he lines questions (untouched by this PR): 
```
    // Finish adding the tracks -- rollback the transaction if the scan did not
    // finish cleanly and the user did not cancel the transaction.
    m_trackDao.addTracksFinish(!m_scannerGlobal->shouldCancel() &&
                               !bScanFinishedCleanly); 
```
tracks from a manual cancelled scan are still added to the library.  

The cleanup tasks are already skipped in 1.12 by the following expression, when you cancel the can during the folder scan phase. 
```
if (bScanFinishedCleanly) {
```

You see that the result should be exactly the same as if we cancel just after the final folder is scanned. Can you confirm this? 

> Your solution might fix the long shutdown time of mixxx in canceled scans but I don't think some one would rescan his entire library during a performance. The better solution in the long run is to improve the performance of this.

Remember that Mixxx automatically starts a rescan in some cases. 
And Yes, it is stupid to do a scan during a performance and once someone notice it he should be able to mediately cancel it. I was myself that stupid, because I could not find a hardly needed track.

An improved scanner performance is only an addition to a cancel function that works almost mediately. 
Even if we manage to reduce the clean up time to 10 % (which will be unlikely possible) we are still talking about 10 long second on a recent device which IMHO require a cancel. 

> It should also be possible to write a unit-test for the clean-up functions. They only depend on internal library state.

Which test cases do you have in mind? What should be verified?

Comment (kain88-de) [inline src/library/dao/trackdao.cpp]: It should be possible to write a MOCK `TrackInfoObject` class that we can use to fill a library in unit-tests. This should allow us to create arbitrary library states. Ideally we could have a function that takes a csv formated string like this.

```csv
Title, Location, Arists
foo, /bar/foo.mp3, blub
bar, /foo/bar.mp3, blub
foo, /foo/foo.mp3, blub
bar, /bar/bar.mp3, blub
```

This defines the state of the library accurately and we can check if these functions do what they are supposed to do. But that would be for another PR.

**Comment by daschuer** [comment]
Thank you for review!
